### PR TITLE
PO to GMP Migration Tool: Podmonitor Auth, TLS and Proxy Migration

### PR DESCRIPTION
**Pull request by karthunni**
Adds migration logic to map Authentication, TLS, and Proxy configurations from Prometheus Operator `PodMonitor` endpoints into GMP `PodMonitoring` equivalents.

This change introduces the core extraction helpers required to consolidate nested Kubernetes `corev1.SecretKeySelector` and `corev1.ConfigMapKeySelector` references into GMP's `monitoringv1.SecretKeySelector` struct. 

**Key Changes:**
* Added translation mapping for `BasicAuth`, `BearerTokenSecret`, `TLSConfig`, and `OAuth2` fields within endpoint configurations.
* Added functions (`extractSecretKey`, `extractConfigMapKey`) that inject clear `<MISSING_...>` placeholders and emit warnings if a referenced resource is missing from the input files.
* Added `convertConfigMapToSecretSelector` to automatically migrate TLS ConfigMap references into generated Secrets, adhering to GMP's strict Secret-only TLS requirements.
* Added comprehensive, table-driven unit tests for all edge cases (cache hits vs misses, empty inputs) in `helpers_test.go`.
* Added a test case to `podmonitor_test.go`.

### CODE REVIEW

File: pkg/migrate/helpers.go [hunk 1]
```diff
@@ -70,3 +76,286 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+	cache *ResourceCache
+	// namespace is the source namespace of the primary resource.
+	namespace string
+	// generatedSecrets accumulates created Secrets when migrating ConfigMaps.
+	generatedSecrets []*unstructured.Unstructured
+}
+
+// extractSecretKey extracts a string value from a Secret, returning a placeholder and warning if not found.
+func extractSecretKey(convCtx *conversionContext, sel corev1.SecretKeySelector) string {
+	obj, ok := convCtx.cache.Get("Secret", convCtx.namespace, sel.Name)
```
[the remark this answers is not in the document]
Comment (karthunni) [reply pkg/migrate/helpers.go]: I instead think it's better to log a WARNING and inject a placeholder string. If we return an error the whole PodMonitor migration will fail, and an unspecified selector name shouldn't be considered a fatal error, the user can easily add it before applying

File: pkg/migrate/helpers.go [hunk 2]
```diff
@@ -70,3 +76,286 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+		}
+		binaryData, found, _ := unstructured.NestedMap(obj.Object, "binaryData")
+		if found {
+			_ = unstructured.SetNestedMap(newSecret.Object, binaryData, "data")
+		}
+		convCtx.generatedSecrets = append(convCtx.generatedSecrets, newSecret)
+	}
+
+	return &monitoringv1.SecretSelector{Secret: &monitoringv1.SecretKeySelector{Name: secretName, Key: secretKey}}
+}
```
[the remark this answers is not in the document]
Comment (karthunni) [reply pkg/migrate/helpers.go]: I believe it would be better to log a WARNING and explicitly mention the Secret won't be generated.

We now also always explicitly set the generated Secret namespace to the input resource namespace

File: pkg/migrate/helpers.go [hunk 3]
```diff
@@ -70,3 +76,286 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+// conversionContext groups common parameters passed down to conversion helper functions.
+type conversionContext struct {
+	logger *slog.Logger
+	// cache provides access to dependent resources.
+	cache *ResourceCache
+	// namespace is the source namespace of the primary resource.
+	namespace string
+	// generatedSecrets accumulates created Secrets when migrating ConfigMaps.
+	generatedSecrets []*unstructured.Unstructured
+}
```
[the remark this answers is not in the document]
Comment (karthunni) [reply pkg/migrate/helpers.go:109]: Instead we will just always explicitly set the generated Secret namespace to the input resource namespace

File: pkg/migrate/helpers.go [hunk 4]
```diff
@@ -70,3 +76,286 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+	}
+	clientID := ""
+	if oa.ClientID.Secret != nil {
+		clientID = extractSecretKey(convCtx, *oa.ClientID.Secret)
+	} else if oa.ClientID.ConfigMap != nil {
+		clientID = extractConfigMapKey(convCtx, *oa.ClientID.ConfigMap)
+	} else {
+		convCtx.logger.Warn("OAuth2 clientID neither defined as Secret nor ConfigMap. Hardcoding placeholder.")
+		clientID = "<MISSING_OAUTH2_CLIENT_ID>"
+	}
```
[the remark this answers is not in the document]
Comment (karthunni) [reply pkg/migrate/helpers.go]: https://github.com/GoogleCloudPlatform/prometheus-engine/pull/1982#discussion_r3539694994

File: pkg/migrate/helpers.go [hunk 5]
```diff
@@ -70,3 +77,345 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+	if tls.CA.Secret != nil || tls.CA.ConfigMap != nil {
+		gmpTLS.CA = convertSecretOrConfigMapToSecretSelector(convCtx, tls.CA)
+	}
+	if tls.Cert.Secret != nil || tls.Cert.ConfigMap != nil {
+		gmpTLS.Cert = convertSecretOrConfigMapToSecretSelector(convCtx, tls.Cert)
+	}
+	if tls.KeySecret != nil {
+		gmpTLS.Key = convertSecretSelector(convCtx, tls.KeySecret)
+	}
```
[the remark this answers is not in the document]
Comment (karthunni) [reply pkg/migrate/helpers.go:334]: Prometheus Operator version currently used in tool does not contain these fields

File: pkg/migrate/helpers.go [hunk 6]
```diff
@@ -70,3 +77,345 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+		return fmt.Sprintf("<MISSING_CONFIGMAP_%s_KEY_%s>", sel.Name, sel.Key)
+	}
+
+	val, found, _ := unstructured.NestedString(obj.Object, "data", sel.Key)
+	if found {
+		return val
+	}
+
+	convCtx.logger.Warn(fmt.Sprintf("Key %q not found in configmap %q. Hardcoding placeholder.", sel.Key, sel.Name))
+	return fmt.Sprintf("<MISSING_KEY_%s_IN_CONFIGMAP_%s>", sel.Key, sel.Name)
```
[the remark this answers is not in the document]
Comment (karthunni) [reply pkg/migrate/helpers.go]: Added handling for binaryData but will not return error, instead log warning and add placeholder

File: pkg/migrate/helpers.go [hunk 7]
```diff
@@ -70,3 +77,345 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+		convCtx.logger.Warn("OAuth2 clientID neither defined as Secret nor ConfigMap. Hardcoding placeholder.")
+		clientID = "<MISSING_OAUTH2_CLIENT_ID>"
+	}
+
+	return &monitoringv1.OAuth2{
+		ClientID:     clientID,
+		ClientSecret: convertSecretSelector(convCtx, &oa.ClientSecret),
+		TokenURL:     oa.TokenURL,
+		Scopes:       oa.Scopes,
+	}
```
[the remark this answers is not in the document]
Comment (karthunni) [reply pkg/migrate/helpers.go]: EndpointParams added. 
Prometheus Operator TLSConfig under OAuth2 does not exist on the version we use in this tool

File: pkg/migrate/helpers.go [hunk 8]
```diff
@@ -70,3 +77,355 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+			if found {
+				_ = unstructured.SetNestedMap(newSecret.Object, data, "stringData")
+			}
+			binaryData, found, _ := unstructured.NestedMap(obj.Object, "binaryData")
+			if found {
+				_ = unstructured.SetNestedMap(newSecret.Object, binaryData, "data")
+			}
+			convCtx.generatedSecrets[secretName] = newSecret
+		}
+	}
```
[the remark this answers is not in the document]
Comment (karthunni) [reply pkg/migrate/helpers.go]: Valid issue, but adding to the resource cache doesn't seem like a great idea, as its purpose is to hold input resources;

Instead we''ve changed to an outputMap at the Migrator level to detect collisions and add a special case for secrets

File: pkg/migrate/helpers.go [hunk 9]
```diff
@@ -70,3 +77,355 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+		if found {
+			return val
+		}
+	}
+
+	// Check standard data field (plain string for ConfigMap, base64 for Secret).
+	val, found, _ := unstructured.NestedString(obj.Object, "data", key)
+	if found {
+		if kind == KindSecret {
+			decoded, err := base64.StdEncoding.DecodeString(strings.TrimSpace(val))
```
[the remark this answers is not in the document]
Comment (karthunni) [reply pkg/migrate/helpers.go:147]: [DecodeString](https://pkg.go.dev/encoding/base64#Encoding.DecodeString) ignores `\r` and `\n`, and I believe `\t` and ` '  '` are invalid

File: pkg/migrate/helpers.go [hunk 10]
```diff
@@ -70,3 +77,355 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+		secrets = append(secrets, c.generatedSecrets[name])
+	}
+	return secrets
+}
+
+// extractResourceKey is a consolidated helper that fetches a key from a ConfigMap or Secret.
+func extractResourceKey(convCtx *conversionContext, kind, name, key string) string {
+	kindUpper := strings.ToUpper(kind)
+	if name == "" {
+		convCtx.logger.Warn(fmt.Sprintf("%sKeySelector has empty name for key %q. Hardcoding placeholder.", kind, key))
```
Comment (bernot-dev) [inline pkg/migrate/helpers.go]: Using slog, these should usually be something like this:
```
convCtx.logger.Warn("Failed to decode base64 data. Hardcoding placeholder.",
    "key", key,
    "kind", kind,
)
```

It rarely makes sense to use fmt.Sprintf inside a logging function. In unstructured logging, there's usually equivalent formatting function available from the logger itself. In structured logging, dividing these parameters into key/value pairs allows warnings/errors to be more easily queried and analyzed.

File: pkg/migrate/helpers.go [hunk 11]
```diff
@@ -70,3 +77,355 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+		return fmt.Sprintf("<MISSING_%s_NAME_KEY_%s>", kindUpper, key)
```
Comment (bernot-dev) [inline pkg/migrate/helpers.go]: It seems like this should return an error at this level, rather than a sentinel pattern value.

File: pkg/migrate/helpers.go [hunk 12]
```diff
@@ -70,3 +77,355 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+		gmpTLS.Cert = convertSecretOrConfigMapToSecretSelector(convCtx, tls.Cert)
+	}
+	if tls.KeySecret != nil {
+		gmpTLS.Key = convertSecretSelector(convCtx, tls.KeySecret)
+	}
+	return gmpTLS
+}
+
+// convertOAuth2 maps PO OAuth2 to GMP OAuth2, extracting the clientID string.
+func convertOAuth2(convCtx *conversionContext, oa *pomonitoringv1.OAuth2) *monitoringv1.OAuth2 {
```
Comment (bernot-dev) [inline pkg/migrate/helpers.go]: It seems like all these functions that take a `*conversionContext` would be better written as receiver methods.

File: pkg/migrate/helpers.go [hunk 13]
```diff
@@ -70,3 +77,355 @@ func ParseAndCleanNamespaces(namespaces []string) []string {
...
+			action = "replace"
+		}
+
+		if action == "labelmap" {
+			logger.Warn("metricRelabelings rule uses 'action: labelmap' which is not supported by GMP and has been dropped.")
+			continue
+		}
+
+		targetLabel := config.TargetLabel
+		if action == "replace" || action == "hashmod" || action == "lowercase" || action == "uppercase" {
```
Comment (bernot-dev) [inline pkg/migrate/helpers.go]: The handlers for the various actions would be much more readable as a switch statement.

File: pkg/migrate/helpers.go (hunk 11 above)
Comment (karthunni) [reply pkg/migrate/helpers.go]: On errors we currently fail the migration, my thought was that we might not want to fail the whole PodMonitor migration if the referenced Secret has an issue. But, we could if that makes more sense.

Also in PO, the [name field is technically optional](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.31/#secretkeyselector-v1-core) (though most likely not correct)

**Review by bernot-dev — CHANGES_REQUESTED** [review]

File: pkg/migrate/helpers.go (hunk 11 above)
Comment (karthunni) [reply pkg/migrate/helpers.go]: Changed to return errors now

**Review by bernot-dev — APPROVED** [review]
LGTM
